ProjFS.Mac/PrjFSKext/public/Message.h

```diff
@@ -47,17 +47,26 @@ struct MessageHeader
     int32_t             pid;
     char                procname[MAXCOMLEN + 1];
 
-    // Size of the flexible-length, nul-terminated path following the message body, including the nul character.
-    uint16_t            pathSizeBytes;
+    // Sizes of the flexible-length, nul-terminated paths following the message body.
+    // Sizes include the nul characters but can be 0 to indicate total absence.
+
+    // Current/destination path
+    uint16_t pathSizeBytes;
+    // Original location (rename, hard link)
+    uint16_t fromPathSizeBytes;
 };
 
 // Description of a decomposed, in-memory message header plus variable length string field
 struct Message
 {
     const MessageHeader* messageHeader;
-    const char*    path;
+
+    const char* path;
+    const char* fromPath;
 };
 
+#if defined(KERNEL) || defined(KEXT_UNIT_TESTING)
```

Comment:
Can you remind of why we need this defined check?

Reply:
We don't define `Message_Init()` in user space, and this file is shared between user space and kernel. We should actually probably just move the kernel-only function prototypes to a kernel-only file instead.